### DISCUSSION TITLE
Extended Font character support?

### DISCUSSION THREAD

**Discussion by gnbl** [russhughes/st7789_mpy#116] [General]
The following does not display properly:

```
>>> import vga2_8x8
>>> display.text(vga2_8x8, "AOU ÄÖÜ äöüß áàâ", 0, y, st7789.RED)
```

Neither does 

`>>> display.text(vga2_8x8, b"AOU ÄÖÜ äöüß áàâ".decode('ascii'), 0, y, st7789.RED)`

What's wrong?

**Comment by russhughes** [comment]
The `text` method uses the old IBM PC BIOS Character fonts that do not contain many extended characters. The `write` method uses fonts converted from TrueType fonts using the font2bitmap.py program in utils. The `write` method supports UTF-8 and extended characters. For an example see: https://github.com/russhughes/st7789_mpy/blob/4d7f2b60d60b739325502285987c3830814d7688/examples/proverbs/proverbs.py and https://github.com/russhughes/st7789_mpy/blob/4d7f2b60d60b739325502285987c3830814d7688/examples/proverbs/proverbs_font.py

**Comment by gnbl** [comment]
But most are in there: https://raw.githubusercontent.com/russhughes/st7789_mpy/master/docs/vga2_8x8.png
I guess my question is how to convert a string literal (in the source code's default UTF-8) to an ASCII coded string that `text()` can handle to use the available glyphs.
Or is my assumption wrong and these fonts do not represent ASCII?

**Comment by gnbl** [comment]
OK so ASCII is only 7 bits. Latin-1, then.

```
>>> ord('ä')
228
>>> chr(228)
'ä'
>>> 'ä'
'ä'
>>> 'ä'.encode()
b'\xc3\xa4'
>>> chr(0xe4)
'ä'
>>> b'\xe4'
b'\xe4'
>>> b'\xe4'.decode('latin1')
'ä'
>>> b'\xe4'.decode('latin-1')
'ä'
>>> 'ä'.encode('latin1')
b'\xe4'
```

This is Python on my computer, I'll give it another try on the ESP32 in about 12 hrs.

**Comment by gnbl** [comment]
https://en.wikipedia.org/wiki/Code_page_437
Python 3.10.6 on linux:
```
>>> 'äöü'.encode('cp437')
b'\x84\x94\x81'
```
Micropython:
```
>>> "äöü".encode('cp437')
b'\xc3\xa4\xc3\xb6\xc3\xbc'
```
It should raise an error instead of ignoring the `encoding` parameter completely.

Regardless, I know what the problem is now, thanks.
